### PR TITLE
discoverd/client: Retry registering services and creating heartbeater

### LINKED ISSUE #2910: flynn bootstrap restore gets unreachable hosts error
When trying to bootstrap restore I get an issue about unreachable hosts. Here are the related gists:
https://gist.github.com/72adfab793f4583bf810b35fc09d0f95
https://gist.github.com/tianx2/eefa84104efffebeac3d32e689e025b3

Had this issue with version 20160606 through version 08

CC @lmars

### LINKED ISSUE #3059: discoverd/client: Service/instance registration should retry automatically
When registering a service/instance we should retry automatically.
This behavior is implemented in multiple places already and should be consolidated within the client as it's effectively required behavior.

Related: #2910

### PR DESCRIPTION
**Pull request by josephglanville**
Closes #3059
Closes #2910

### CODE REVIEW

**Comment by josephglanville** [comment]
I tried a few other approaches but this somewhat obtuse approach seemed to be the most robust.
This is mostly useful during cluster states where the scheduler is unable to restart jobs that exit.
Specifically during bootstrap before the scheduler has been initialised.

File: discoverd/client/heartbeat.go [hunk 1]
```diff
@@ -73,11 +83,18 @@ func (c *Client) RegisterInstance(service string, inst *Instance) (Heartbeater,
 		inst.Meta[kv[0]] = kv[1]
 	}
 	h := newHeartbeater(c, service, inst)
-	firstErr := make(chan error)
-	go h.run(firstErr)
-	if err := <-firstErr; err != nil {
+	err := runAttempts.Run(func() error {
+		firstErr := make(chan error)
+		go h.run(firstErr)
+		if err := <-firstErr; err != nil {
```
Comment (titanous) [inline discoverd/client/heartbeat.go]: I think this can just be `return <-firstErr`

**Comment by titanous** [comment]
LGTM pending comment

**Comment by lmars** [comment]
LGTM

**Pull request merged**